pkg/morph/client/notifications.go

```diff
@@ -214,72 +234,109 @@ func (c *Client) restoreSubscriptions(cli *rpcclient.WSClient, endpoint string)
 	stopCh := make(chan struct{})
 	defer close(stopCh)
 
+	blockRcv := make(chan *block.Block)
+	notificationRcv := make(chan *state.ContainedNotificationEvent)
+	notaryReqRcv := make(chan *result.NotaryRequestEvent)
+
 	// neo-go WS client says to _always_ read notifications
 	// from its channel. Subscribing to any notification
 	// while not reading them in another goroutine may
 	// lead to a dead-lock, thus that async side notification
 	// listening while restoring subscriptions
 	go func() {
+		var e any
+		var ok bool
+
 		for {
 			select {
 			case <-stopCh:
 				return
-			case n, ok := <-cli.Notifications:
-				if !ok {
-					return
-				}
+			case e, ok = <-blockRcv:
+			case e, ok = <-notificationRcv:
+			case e, ok = <-notaryReqRcv:
+			}
+
+			if !ok {
+				return
+			}
 
-				c.notifications <- n
+			if background {
+				// background client (test) switch, no need to send
+				// any notification, just preventing dead-lock
+				continue
 			}
+
+			c.routeEvent(e)
 		}
 	}()
 
+	if background {
```

Comment:
If it is false, do can we take a mutex too?
It could look more readable without `if`.

Reply:
no, we would have a dead-lock then: `background` flag is `false` only if we have already lost connection and try to establish a new one, so `switchLock` has been already taken